Add initramfs support for clevis. v3

Phant0mas opened the pull request:
This pull request consists of ShaRose's work rebased in one commit, a commit adding TPM2 support and instructions on how to generate a new working initramfs.

These changes where tested on Ubuntu 16.04, kernel version 4.14 with Luks encryption using TPM2.

Original pull request: https://github.com/latchset/clevis/pull/18

martinezjavier commented:
@Phant0mas thanks for this, a couple of general comments:

* I think initramfs-tools will be a better naming than just initramfs (since we do support initramfs generated by dracut). So please change `src/initramfs/` to `src/initramfs-tools` and so on.

* Shouldn't we have a `Makefile.am` in `src/initramfs-tools` so the hooks and scripts are installed under  are installed under `/usr/share/initramfs-tools` or `/etc/initramfs-tools` ? Probably we also need a configure option to choose to build with dracut or initramfs-tools support and also another option to choose the path where the initramfs-tools hook and scripts will be installed (like we have for `--with-dracutmodulesdir`).

martinezjavier commented on src/initramfs/hooks/clevis:
```diff
@@ -27,6 +27,10 @@ copy_exec /usr/local/bin/clevis-decrypt-http
+copy_exec /usr/local/bin/clevis-decrypt-tpm2
+copy_exec /usr/local/bin/tpm2_createprimary
+copy_exec /usr/local/bin/tpm2_unseal
+copy_exec /usr/local/bin/tpm2_load
```
I forgot to mention this before, but we should check if the tpm2-tools are installed in the system before attempting to copy these. Please see the following fixup commit for the dracut module setup as an example:

https://github.com/latchset/clevis/commit/318eaade57fdc5588a6b9e83ed01ede12c5f62c7

Phant0mas replied:
I can't figure out autotools for the life of me. 
Inside configure.ac I have added the following
AC_ARG_ENABLE([initramfstools],
    AS_HELP_STRING([--enable-initramfstools], [Enable initramfs-tools]))

AS_IF([test "x$enable_initramfstools" = "xyes"], [
    AC_ARG_WITH([initramfstoolsmodulesdir],
            [AS_HELP_STRING([--with-initramfstoolsmodulesdir=DIR], [Directory for initramfs-tools modules])],
            [],
            [with_initramfstoolsmodulesdir=$initramfstoolsmodulesdir])
    AC_SUBST([initramfstoolsmodulesdir], [$with_initramfstoolsmodulesdir])
])

For example with this `./configure --enable-initramfstools=xyes --with-initramfstoolsmodulesdir=/tmp/test/` the var $initramfstoolsmodulesdir should be /tmp/test.

In the Makefile.am inside initramfs, I have added:
initramfstoolsdir = @initramfstoolsmodulesdir@/60$(PACKAGE_NAME)
nodist_initramfstools_SCRIPTS = hooks/clevis scripts/local-bottom/clevis \
	scripts/local-top/clevis
EXTRA_DIST=hooks/clevis.in scripts/local-bottom/clevis.in scripts/local-top/clevis.in
CLEANFILES=hooks/clevis scripts/local-bottom/clevis scripts/local-top/clevis

%: %.in
	$(AM_V_GEN)mkdir -p $(dir $@)
	$(AM_V_GEN)$(SED) \
	-e 's,@initramfstoolsdir\@,$(initramfstoolsdir),g' \
	$(srcdir)/$@.in > $@

which means that @initramfstoolsdir@ should change into /tmp/test when I run make.
It doesn't. Instead it changes into ". /60clevis"

Any ideas?

martinezjavier replied:
@Phant0mas you don't have to use `./configure --enable-initramfstools=xyes` but `./configure --enable-initramfstools=yes`. The `"x$foo = xyes" check is a portable way to test since in some shells testing for an empty value is not that easy.

But in your example it will be `xxyes = xyes` so the condition will be false.

martinezjavier commented on INSTALL.md:
```diff
@@ -78,3 +78,12 @@ initramfs with:
+sudo cp -rv src/initramfs/* /usr/share/initramfs-tools/
+sudo chmod +x /usr/share/initramfs-tools/scripts/hooks/clevis
+sudo chmod +x /usr/share/initramfs-tools/scripts/local-top/clevis
+sudo chmod +x /usr/share/initramfs-tools/scripts/local-bottom/clevis
```
If you hook this with autotools as suggested, then there shouldn't be a need for these instructions.

martinezjavier commented on INSTALL.md:
```diff
@@ -78,3 +78,12 @@ initramfs with:
+sudo chmod +x /usr/share/initramfs-tools/scripts/hooks/clevis
+sudo chmod +x /usr/share/initramfs-tools/scripts/local-top/clevis
+sudo chmod +x /usr/share/initramfs-tools/scripts/local-bottom/clevis
+sudo update-initramfs -u -k 'all'
```
This will still be needed, but you can add this into the [README.md](https://github.com/latchset/clevis/blob/master/README.md), under a new *Unlocker: Initramfs-tools* section.

Phant0mas commented:
Because of the nature of the changes, I chose not to create more commits but instead rebase the changes on the old commits.
What I did:
1) Renamed src/initramfs to src/initramfs-tools
2) Added option to choose which of the two will be enabled (Dracut or initramfs).
3) Added option to choose when initramfs hooks will be installed.
4) Added makefile.am inside src/initramfs-tools.
5) Added check to only copy the tpm2-tools only if they exist.

Please check my changes and report any issues :)

martinezjavier commented on src/initramfs-tools/hooks/clevis.in:
```diff
@@ -27,6 +27,12 @@ copy_exec /usr/local/bin/clevis-decrypt-http
 copy_exec /usr/local/bin/clevis-decrypt-tang
 copy_exec /usr/local/bin/clevis-decrypt-sss
 copy_exec /usr/local/bin/clevis-decrypt
+if [ -x /usr/local/bin/clevis-decrypt-tpm2 ]; then
```
I'll check for all binaries in a for loop. The fact that clevis-decrypt-tpm2 was installed just means that the tpm2-tools were available at installation time.

But what happens if a user later removes the tpm2-tools package? This may or may not be a hard dependency on the system so it's something to take into account.

Phant0mas replied:
Will this be an acceptable solution?
```
if [ -x @bindir@/clevis-decrypt-tpm2 ]; then
    copy_exec @bindir@/clevis-decrypt-tpm2
fi
if [ -x @bindir@/tpm2_createprimary ]; then
    copy_exec @bindir@/tpm2_createprimary
    copy_exec @bindir@/tpm2_unseal
    copy_exec @bindir@/tpm2_load
fi
```
The tpm2_ binaries are all part of tpm2-tools. So I think we can assume that if one of them is present then the other must be as well.

martinezjavier replied:
@Phant0mas yes, I know. I'm just not sure about testing for a single binary as the heuristic to know if all the other are available. But yes, in practice I think that testing just for one is enough.

martinezjavier commented on src/initramfs-tools/hooks/clevis.in:
```diff
@@ -0,0 +1,34 @@
+
+. @initramfstoolsdir@/hook-functions
+
+copy_exec /usr/local/bin/clevis-decrypt-http
```
Sorry for missing this in the previous version, but the absolute path shouldn't be hardcoded here. Instead just either only have the tools without an absolute path (and rely on `$PATH`) or use `@bindir@/clevis-decrypt-http`.

I personally prefer the latter but I see that the dracut module does the former, so I don't have a strong opinion on this.

Phant0mas replied:
Should I also use something similar for 
```
copy_exec /usr/bin/curl
copy_exec /bin/bash
```

martinezjavier replied:
@Phant0mas probably yes. Although I'm now leading to just leave the commands and rely on `$PATH`. I guess `copy_exec` / initramfs-tool is able to handle that?

Phant0mas replied:
Using only the patch makes update-initramfs to fail:
``` 
$ sudo update-initramfs -u -k 'all'
update-initramfs: Generating /boot/initrd.img-4.14.0-041400-generic
E: /usr/share/initramfs-tools/hooks/clevis failed with return 1.
update-initramfs: failed for /boot/initrd.img-4.14.0-041400-generic with 1.
```

And seeing the other hooks, they are using full paths. Maybe we should keep the @bindir@. WDYT?

martinezjavier replied:
@Phant0mas I see, haven't used initramfs-tools in years. Then yes, I think we should use `@bindir@` then.

Phant0mas replied:
Now I am trying to find which autotool variable points to /usr/bin and /bin. The variable prefix for example points to /usr/local/bin by default and I can't find anything related in the documentation. Any ideas?

npmccallum commented:
Can't we just detect dracut and initramfs and behave accordingly? Why force the builder to specify options?

Phant0mas commented:
Hello @npmccallum @martinezjavier sorry for not responding earlier.
Will something like this do? I tested it and it works. Of course in the example bellow I take for granted that one system will either have dracut or initramfs-tools. 
AC_DEFUN(AC_PROG_DRACUT, [AC_CHECK_PROG([DRACUT], [dracut], [yes])])
AC_PROG_DRACUT
if test x"${DRACUT}" == x"yes" ; then
    AC_ARG_WITH([dracutmodulesdir],
            [AS_HELP_STRING([--with-dracutmodulesdir=DIR], [Directory for dracut modules])],
            [],
            [with_dracutmodulesdir=$($PKG_CONFIG --variable=dracutmodulesdir dracut)])
    AC_SUBST([dracutmodulesdir], [$with_dracutmodulesdir])
else
    AC_ARG_WITH([initramfstoolsmodulesdir],
            [AS_HELP_STRING([--with-initramfstoolsmodulesdir=DIR], [Directory for initramfs-tools modules])],
            [],
            [with_initramfstoolsmodulesdir=$initramfstoolsmodulesdir])
    AC_SUBST([initramfstoolsmodulesdir], [$with_initramfstoolsmodulesdir])
fi

circhioz commented:
@npmcallum @Phant0mas I use clevis on Arch Linux which doesn't use either dracut or initramfs-tools. I think it would be better if autoconf detected this situation instead of assuming we have initramfs-tools if it doesn't find dracut

Phant0mas commented:
@circhioz @npmccallum Then I will also need to add a case for whether initramfs exists or else to fail. Btw if you use it with mkinitcpio it would be great if someone adds support for it as well.

Phant0mas commented:
@npmccallum @circhioz If either dracut or update-initramfs is missing it will fail. In the old version if dracut was missing it failed.

npmccallum commented on src/initramfs-tools/hooks/clevis.in:
```diff
@@ -0,0 +1,34 @@
+copy_exec @bindir@/clevis-decrypt-http
+copy_exec @bindir@/clevis-decrypt-tang
+copy_exec @bindir@/clevis-decrypt-sss
+copy_exec @bindir@/clevis-decrypt
```
We appear to be missing the tpm2 pin and its utilities here. I suspect this just needs to be updated given the tpm2 support that landed recently.

Phant0mas replied:
@npmccallum I am sorry but I am not sure I understand what you mean. Directly bellow, in the commit be1a56bd8f8b5d51b4826b18925226c1d3cb01ce "Update hook to include tpm2 pin." I add:
```
+if [ -x @bindir@/clevis-decrypt-tpm2 ]; then
+    copy_exec @bindir@/clevis-decrypt-tpm2
+fi
+if [ -x @bindir@/tpm2_createprimary ]; then
+    copy_exec @bindir@/tpm2_createprimary
+    copy_exec @bindir@/tpm2_unseal
+    copy_exec @bindir@/tpm2_load
+fi
```
In case it doesn't detect clevis-decrypt-tpm2 and tpm2_createprimary, it doesn't copy clevis-decrypt-tpm2 and the tpm2-tools we need. Did I miss something?

martinezjavier replied:
@npmccallum @Phant0mas Github isn't great at this and the comment is shown in both the first and third commit.

The reason why the tpm2 commands are not here is that @Phant0mas based his work on a previous pull request that predates the tpm2 pin, so he added the tpm2 bits in a following commit.

npmccallum commented:
Mea culpa. @Phant0mas can you make sure that this series of patches is based on logical changes and not stream of development? That makes it much easier to review. I'd like to pull this in for v11, so I'd like to wrap up the review soon.

Phant0mas commented:
@npmccallum I could merge the first 2 patches which change code and add initramfs-tools support, into one and leave the documentation change as the other commit. WDYT?

blaufish commented:
@Phant0mas @npmccallum 1/ is this PR still under development or abandoned? 2/ does this work better with Ubuntu16, 18 than the dracut unlocker?

Phant0mas commented:
Yes it is still under development. I was waiting for @npmccallum to tell me if it was a good idea to merge the 3 commits into one.

blaufish commented on configure.ac:
```diff
@@ -17,18 +17,36 @@ PKG_CHECK_MODULES([jansson], [jansson >= 2.10])
-PKG_CHECK_MODULES([dracut], [dracut])
 PKG_CHECK_MODULES([audit], [audit >= 2.7.8])
 
+AC_DEFUN(AC_PROG_DRACUT, [AC_CHECK_PROG([DRACUT], [dracut], [yes])])
```
AC_DEFUN([AC_PROG_DRACUT], [AC_CHECK_PROG([DRACUT], [dracut], [yes])])
AC_DEFUN([AC_PROG_INITRAMFS_TOOLS], [AC_CHECK_PROG([INITRAMFS_TOOLS], [update-initramfs], [yes])])

Phant0mas replied:
Thank you! After work today I will fix it.

npmccallum commented:
@Phant0mas It would be good to either merge these into one patch or into a set of smaller logical changes. I don't see an obvious place for the latter, so the former works for me. But if you see a good place to break it up, then go ahead.

Phant0mas commented:
@npmccallum okay I followed your suggestion merged everything into one, added @blaufish 's fix and pushed the updated commit.

De3Daniel commented:
@Phant0mas When using this code as is, if the network Ethernet cable was unplugged it took 6 minutes before I could type in my password. This is something I used to bypass the wait time in the case of no cable. 

-[ "$boot" = nfs ] || configure_networking
+eth_check() {
+    for device in $(all_netbootable_devices); do
+        ip link set dev $device up
+        sleep 1
+        ETH_HAS_CARRIER=$(</sys/class/net/$device/carrier)
+        if [ "$ETH_HAS_CARRIER" == '1' ]; then
+             return 0
+            
+        fi    
+  done
+
+  return 1
+}
+
+if eth_check; then
+    [ "$boot" = nfs ] || configure_networking
+    
+else
+   echo No Network Connection
+fi

De3Daniel commented:
@Phant0mas  When using this code as is, if the network Ethernet cable was unplugged it took 6 minutes before I could type in my password. This is something I used to bypass the wait time in the case of no cable. 

# Make sure networking is set up: if booting via nfs, it already is
 # Doesn't seem to work when added to clevisloop for some reason
-[ "$boot" = nfs ] || configure_networking
+eth_check() {
+    for device in $(all_netbootable_devices); do
+        ip link set dev $device up
+        sleep 1
+        ETH_HAS_CARRIER=$(</sys/class/net/$device/carrier)
+        if [ "$ETH_HAS_CARRIER" == '1' ]; then
+             return 0      
+        fi    
+  done
+  return 1
+}
+if eth_check; then
+    [ "$boot" = nfs ] || configure_networking    
+else
+   echo No Network Connection
+fi

Phant0mas commented:
@De3Daniel awesome, thank you! I will add it and create an update.

Phant0mas commented:
Hello @De3Daniel, I just got around to try out your suggestion and I think the best place to add this is inside src/initramfs-tools/scripts/local-top/clevis.in and not src/initramfs/hooks/clevis.in. WDYT?

```
--- a/src/initramfs-tools/scripts/local-top/clevis.in
+++ b/src/initramfs-tools/scripts/local-top/clevis.in
@@ -110,9 +110,25 @@ clevisloop()
 
 . /scripts/functions
 
-# Make sure networking is set up: if booting via nfs, it already is
-# Doesn't seem to work when added to clevisloop for some reason
-[ "$boot" = nfs ] || configure_networking
+# Check if network is up before trying to configure it.
+eth_check() {
+    for device in $(all_netbootable_devices); do
+        ip link set dev $device up
+        sleep 1
+        ETH_HAS_CARRIER=$(</sys/class/net/$device/carrier)
+        if [ "$ETH_HAS_CARRIER" == '1' ]; then
+            return 0
+        fi
+    done
+    return 1
+}
+if eth_check; then
+    # Make sure networking is set up: if booting via nfs, it already is
+    # Doesn't seem to work when added to clevisloop for some reason
+    [ "$boot" = nfs ] || configure_networking
+else
+    echo No Network Connection
+fi

```

De3Daniel commented:
I agree, I miss quoted the location. It should be in 'src/initramfs-tools/scripts/local-top/clevis.in'

Phant0mas commented:
I updated the pull request with everyone's suggestions. Is it ready now to be merged?

dguglielmi commented on src/initramfs-tools/Makefile.am:
```diff
@@ -0,0 +1,18 @@
+		-e 's,@bindir\@,$(bindir),g' \
+		$(srcdir)/$@.in > $@
+
+install-nodist_initramfstoolsSCRIPTS:
```
There is a typo here ?

install-nodist_initramfstoolsSCRIPTS:

should be 

install-nodist_initramfstools_SCRIPTS:

?

michaelbarkdoll replied:
@dguglielmi hi, i have same issue when trying to build the .deb.  How did you fix this particular issue?

michaelbarkdoll replied:
Sorry for the noise, I was able to fix
````
sudo dpkg-buildpackage -rfakeroot -b -d
````

by adding:
debian/rules:

dh_auto_configure -- --with-dracutmodulesdir=/usr/lib/dracut/modules.d --with-initramfstoolsmodulesdir=/usr/share/initramfs-tools

XueSheng-GIT commented:
I'm using Ubuntu 18.04. Server with Tang and client with clevis (automatic unlock of luks root).
After adding the initramfs hook and scripts of this PR, the automatic unlock at boot does work as expected. Thanks for your effort!

Unfortunately I'm not able to use multiple tang servers (redudancy) for this setup. Each tang server is bound to a separate key slot. But at boot, only the first clevis slot is used. Thus if the first tang server is not available, the boot process gets stuck. I would expect that if one tang server is not available, the next slot is checked and so on.

Should it be possible to use multiple tang servers with initramfs?

kevhead89 commented:
@XueSheng-GIT i'm testing tang with clevis now on 18.04 but can't seem to get the auto unlock to work. How did you apply this patch to get it to work? I'm only using 1 tang server right now for testing

XueSheng-GIT commented:
@kevhead89 I manually modified the files (according to the changed files of this pull request). If I'm not mistaken, that was all I had to do (of course the networking must be configured to be up on boot).
That's all I can remember. I've switched to mandos because I'm mainly working with ubuntu/debian and debian stable does not provide clevis/tang so far.

Phant0mas commented:
I will update the files to solve the conflicts

mhalano commented:
@Phant0mas Any news about that?

Phant0mas commented:
@mhalano the pull request in its current form cannot work with the upstream clevis, as autotools are no longer used. I need to port this patch to meson.

Phant0mas commented:
@martinezjavier ok I am trying to build clevis with meson and I am getting the following:

```
qadmin@serverpack:~/repos/clevis$ meson build
The Meson build system
Version: 0.29.0
Source dir: /home/qadmin/repos/clevis
Build dir: /home/qadmin/repos/clevis/build
Build type: native build
Build machine cpu family: x86_64
Build machine cpu: x86_64
Project name: clevis
Native c compiler: cc (gcc 5.4.0-6ubuntu1)

Meson encountered an error in file meson.build, line 4, column 0:
Tried to access unknown option "libexecdir".
```

mtottenh commented:
> EDIT: I solved this, the cause was old meson version in the Ubuntu 16.04 I was using.

@Phant0mas - Any update on this PR? 

I took a quick crack at porting your PR to meson it didn't seem too difficult:

https://github.com/mtottenh/clevis/commit/44adba327e0ff5363a2326737ca261bab08825b2

dkopecek commented:
Closing as this seems obsolete. Let's continue with #101.
